src/validation.cpp

```diff
@@ -620,19 +621,70 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
+                // changed mempool entries. Due to both implementation and runtime complexity concerns, this isn't
+                // very realistic, either, thus we we try to ensure a limited set of transaction is RBF'able despite
+                // mempool conflicts here. Importantly, we want to ensure that transactions which were accepted using
+                // the below carve-out are able to be RBF'ed, without impacting the security the carve-out provides
+                // for off-chain contract systems (see link in the comment below).
+                //
+                // Specifically, the subset of RBF transactions which we allow despite conflicts are those which
+                // conflict directly with exactly one other transaction (but may evict children of said transaction),
+                // and which are not adding any new mempool dependencies (checked first).
+                // Such transactions are clearly not merging any existing packages, so we are only concerned with
+                // ensuring that (a) no package is growing past the package size (not count) limits and (b) we are
+                // not allowing something to effectively use the (below) carve-out spot when it shouldn't be allowed
+                // to.
+                // To check these we re-run CalculateMemPoolAncestors up to twice, once corresponding to the above,
+                // regular, case, and one corresponding to the below, carve-out, case. Each has the same ancestor
+                // count limits as the corresponding case (as the ancestor limits should be the same for both our new
+                // transaction and any conflicts, though you shouldn't be able to violate these given the
+                // no-new-unconfirmed-inputs check), with the descendant limits increased by the size (and one
+                // count) of the directly-conflicting transaction (as these are recalculated by assuming the new
+                // transaction being added is a new dependant, with no removals, of each parents' existing
+                // dependant set).
+                assert(setIterConflicting.size() == 1);
+                CTxMemPool::txiter conflict = *setIterConflicting.begin();
+
+                std::set<uint256> conflict_existing_mempool_inputs;
+                for (const CTxMemPool::txiter& txit : pool.GetMemPoolParents(conflict)) {
+                    conflict_existing_mempool_inputs.insert(txit->GetTx().GetHash());
+                }
+
+                for (const CTxIn& new_input: tx.vin) {
+                    if (conflict_existing_mempool_inputs.count(new_input.prevout.hash) == 0) {
+                        if (pool.exists(new_input.prevout.hash)) {
+                            return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_NONSTANDARD, "too-long-mempool-chain", dummy_err_string);
```

Comment:
This check appears to be identical to https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L733

which would at least be good to point out, or how it's different, if it's not the same. I could see the other check being relaxed in the future but this kept in place.

Reply:
Indeed! Good call. I removes the entire block of code wholesale, but added extra comments to ensure if the BIP 125 rules are relaxed we know to re-add the checks here.